
import java.util.Scanner;
import java.io.FileInputStream;


class Solution
{
	static int[] parent;
	
	static int find(int x) {
		if(parent[x] ==x) {
			return x;
		}
		
		return parent[x] = find(parent[x]);
	}
	
	static void union(int a, int b) {
		int rootA = find(a);
		int rootB = find(b);
		
		if(rootA!=rootB) {
			parent[rootB] = rootA;
		}
		
	}
	
	
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		Scanner sc = new Scanner(System.in);
		int T;
		T=sc.nextInt();

		for(int test_case = 1; test_case <= T; test_case++)
		{
			
			int n = sc.nextInt();
			int m = sc.nextInt();
			
			parent = new int[n+1];
			
			for(int i=1; i<=n; i++)
			{
				parent[i] = i;
			}
			
			StringBuilder sb = new StringBuilder();
			
			for(int i=0; i<m; i++)
			{
				int command = sc.nextInt();
				int a = sc.nextInt();
				int b = sc.nextInt();
				
				
				if(command==0) {
					union(a,b);
				}
				
				else {
					if(find(a)==find(b)){
						sb.append("1");
					} else {
						sb.append("0");
					}
				}
				
			}
			
			System.out.println("#" + test_case + " " + sb);
			
			
		}
	}
}